watchdog/watchdog.go

```diff
@@ -39,95 +44,165 @@ type Log interface {
+			w.Log.Infof("watchdog is shutting down, stop heartbeating")
+			return
+		case <-time.After(delay):
+			continue
+		}
 	}
 }
 
 // watchProgram will monitor a program and restart it if it exits.
 // This method will run forever.
-func watchProgram(program Program, restartDelay time.Duration, log Log, exitAll func()) {
+func (w *Watchdog) startProgram(idx int) {
+	program := &(w.Programs[idx])
 	for {
 		start := time.Now()
-		log.Infof("Starting %#v", program)
-		cmd := exec.Command(program.Path, program.Args...)
-		err := cmd.Run()
+		err := program.Run(w.Log, w.shutdownCh)
 		if err != nil {
-			log.Errorf("Error running program: %q; %s", program, err)
+			w.Log.Errorf("Error running %s: %+v; %s", program.Name, program, err)
 		} else {
-			log.Infof("Program finished: %q", program)
+			w.Log.Infof("%s finished: %+v", program.Name, program)
 			if program.ExitOn == ExitOnSuccess {
-				log.Infof("Program configured to exit on success, not restarting")
+				w.Log.Infof("Program %s configured to exit on success, not restarting", program.Name)
 				break
 			} else if program.ExitOn == ExitAllOnSuccess {
-				log.Infof("Program configured to exit on success, exiting")
-				exitAll()
+				w.Log.Infof("Program %s configured to trigger full watchdog shutdown", program.Name)
+				w.Shutdown()
 			}
 		}
-		log.Infof("Program ran for %s", time.Since(start))
-		if time.Since(start) < restartDelay {
-			log.Infof("Waiting %s before trying to start command again", restartDelay)
-			time.Sleep(restartDelay)
+		w.Log.Infof("Program %s ran for %s", program.Name, time.Since(start))
+		select {
+		case <-w.shutdownCh:
+			w.Log.Infof("watchdog is shutting down, not restarting %s", program.Name)
+			return
+		default:
+		}
+		if time.Since(start) < w.RestartDelay {
+			w.Log.Infof("Waiting %s before trying to start %s command again", w.RestartDelay, program.Name)
+			time.Sleep(w.RestartDelay)
+		}
+	}
+}
+
+// terminateExistingMatches aggressively kills anything running that looks like similar
+// to what this watchdog will be running. the goal here is to be sure that, if multiple
+// calls attempt to start a watchdog, the last one will be the only one that survives.
+func (w *Watchdog) terminateExistingMatches() {
+	w.Log.Infof("Terminate any existing programs that look like matches")
+	var killedPids []int
+	for i := 1; i <= 3; i++ {
+		killedPids = w.killSimilarRunningPrograms()
+		if !includesARealProcess(killedPids) {
+			w.Log.Infof("none of these programs are running")
+			return
 		}
+		w.Log.Infof("Terminated pids %v", killedPids)
+		time.Sleep(terminationDelay)
 	}
 }
+
+func includesARealProcess(pids []int) bool {
+	for _, p := range pids {
+		if p > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func (w *Watchdog) killSimilarRunningPrograms() (killedPids []int) {
+	// kill any running processes that look like the ones this watchdog wants to watch
+	// this logic also exists in the updater, so if you want to change it, look there too.
```

Comment:
This code needs to be changed in two places? Or is the comment old?

Reply:
comment is still valid. the updater needs to shut down everything but itself before downloading and kicking off an update. it does that by (1) looking for the watchdog (i.e. the keybase service process which is the parent of another keybase service process) and killing it, then (2) executing similar code to this which kills everything that looks like a match except for its pid.